added CloudFront invalidation feature

martazobro opened the pull request.

martazobro commented:
Hello @ndeloof , bsideup.
We are using s3 as a central resources storage. All files are reached through CDN (CloudFront distributions). It's extremely important of us to invalidate files right after upload, so all caches are updated. 
This is what extension does.
The feature might be helpful for others teams as well.
I'm writing to you, guys, as most recent plugin contributors.
Could you, please, take a look and advise me what should i do next to have this feature as part of plugin.

Thank you in advance!
Have a nice day,
Marta

bsideup commented:
@martazobro at least I see some code style issues. Could you please format it in appropriate way?

bsideup commented:
@martazobro also, I would prefer to see this as an additional action, not a part of main action. So you will have 2 actions: first one for upload and second one (abstract) for CloudFront invalidation. Reason: users can re-use this action to invalidate cloudfront without uploading.

martazobro commented:
Hi, @bsideup
Implemented CloudFront invalidation as separate action, please, take a look.
Thank you in advance!

bsideup commented on src/main/java/hudson/plugins/s3/cloudfront/InvalidationEntry.java:
```diff
@@ -0,0 +1,64 @@
+        }
+        
+    	public FormValidation doCheckBucket(@QueryParameter String bucket) {
+			return checkNotBlank(bucket, "Bucket name must be speified");
```
typo in "speified" :)

bsideup commented on src/main/java/hudson/plugins/s3/cloudfront/InvalidationEntry.java:
```diff
@@ -0,0 +1,64 @@
+		}
+		
+		public FormValidation doCheckKeyPrefix(@QueryParameter String keyPrefix) {
+			return checkNotBlank(keyPrefix, "Key prefix name must be speified");
```
same here

bsideup commented on src/main/resources/hudson/plugins/s3/cloudfront/InvalidationEntry/help-bucket.html:
```diff
@@ -0,0 +1 @@
+<div>Destination bucket. Environment variable can be used, for example</div>
```
missing example?

bsideup commented on src/main/webapp/help-invalidate.html:
```diff
@@ -0,0 +1,3 @@
+<div>Invalidates Amazon S3 assets among available cloudFront distributions.
```
cloudFront -> CloudFront

bsideup commented on src/main/java/hudson/plugins/s3/cloudfront/CloudFrontInvalidatePublisher.java:
```diff
@@ -0,0 +1,188 @@
+	public static S3Profile getProfile(String profileName) {
+		S3Profile[] profiles = DESCRIPTOR.getProfiles();
+
+		if (profileName == null && profiles.length > 0)
```
you can remove this lines and do something like this (in cycle):
```
if (profileName == null || profile.getName.equals(profileName)) {
   return profile;
}
```

bsideup commented on src/main/java/hudson/plugins/s3/cloudfront/CloudFrontInvalidatePublisher.java:
```diff
@@ -0,0 +1,188 @@
+
+				if (entry.noInvalidateOnFailure && buildFailed) {
+					// build failed. don't post
+					log(listener.getLogger(), "Skipping S3 key invalidation because build failed");
```
AFAIK invalidation applied to CloudFront, not S3

martazobro replied:
That's true, i was looking at CloudFront in relation with S3. Actually, that's how we are using it. After we update some resources on S3, we want them to be updated within cdn as well, so we invoke invalidation on those keys. And because of the relation, we ask user to provide s3 bucket name and key prefix.
The callable, extracts all distributions related with the account and invalidate all distributions related to the bucket, so all of them have updated version of S3 file.

On the other hand, we can ask user to specify distribution alias instead of bucket and CloudFront object path prefix. @bsideup , do you think it would be more useful approach? What if user has multiple caches for same entries, would he/she has to create multiple invalidationEntries in the job config with same entry path?

bsideup commented on src/main/java/hudson/plugins/s3/cloudfront/CloudFrontInvalidatePublisher.java:
```diff
@@ -0,0 +1,188 @@
+	@Override
+	public boolean perform(AbstractBuild<?, ?> build, Launcher launcher, BuildListener listener) throws InterruptedException, IOException {
+
+		final boolean buildFailed = build.getResult() == Result.FAILURE;
```
fail fast if true

martazobro replied:
Each entry has checkbox whether to perform invalidation if build failed. It was done the same way as upload to s3, if resources were uploaded then probably they should be invalidated

bsideup commented on src/main/java/hudson/plugins/s3/cloudfront/CloudFrontInvalidatePublisher.java:
```diff
@@ -0,0 +1,188 @@
+		
+		public S3Profile[] getProfiles() {
+			if(profiles.isEmpty()){
+				profiles.addAll(Arrays.asList(((hudson.plugins.s3.S3BucketPublisher.DescriptorImpl) Jenkins.getInstance().getDescriptor(S3BucketPublisher.class)).getProfiles()));
```
this line is too long, could you please extract some parts like getting descriptors?

bsideup commented on src/main/java/hudson/plugins/s3/cloudfront/callable/AbstractCloudFrontCallable.java:
```diff
@@ -0,0 +1,35 @@
+	private final String accessKey;
+	private final Secret secretKey;
+	private final boolean useRole;
+	private AmazonCloudFront cloudFrontClient;
```
you should mark cloudFrontClient transient since it's not serializable (and it shouldn't be serialized anyway)

martazobro commented:
@bsideup , i've fixed your notations, improved invalidation logic by enabling invalidation paths with wildcard and added possibility to specify comma-separated list of paths.
